funcframework/framework.go

```diff
@@ -206,6 +212,64 @@ func wrapEventFunction(fn interface{}) (http.Handler, error) {
 	}), nil
 }
 
+func wrapTypedFunction(fn interface{}) (http.Handler, error) {
+	err := validateTypedFunction(fn)
+	if err != nil {
+		return nil, err
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, err := readHTTPRequestBody(r)
+		if err != nil {
+			writeHTTPErrorResponse(w, http.StatusBadRequest, crashStatus, fmt.Sprintf("%v", err))
+			return
+		}
+		argVal := reflect.New(reflect.TypeOf(fn).In(0))
+
+		if err := json.Unmarshal(body, argVal.Interface()); err != nil {
+			writeHTTPErrorResponse(w, http.StatusBadRequest, crashStatus, fmt.Sprintf("Error: %s, while converting input type data: %s", err.Error(), string(body)))
+			return
+		}
+
+		defer recoverPanic(w, "user function execution")
+		funcReturn := reflect.ValueOf(fn).Call([]reflect.Value{
+			argVal.Elem(),
+		})
+
+		if len(funcReturn) >= 1 {
+			var errorVar error
+			errorType := reflect.TypeOf(&errorVar).Elem()
+			firstVal := funcReturn[0].Interface()
+			if !reflect.TypeOf(firstVal).AssignableTo(errorType) {
+				returnVal, _ := json.Marshal(firstVal)
+				fmt.Fprintf(w, string(returnVal))
```

Comment:
In the single return case (i.e. only an error or only a result value) this LGTM. In the multi-return value case though I'd expect the error value to be checked first and to override the serialization of any result passed back. 

Proposal: simplify this code by checking if the last value is an error first (and also relying on the invariants established by your validation function): if the last value is error type and non-nil, then `writeHTTPErrorResponse` and return. Otherwise, if the first value isn't an error we can just return that.

I think the overall implementation might be simplified to something like
```
var errorVar error
errorType := reflect.TypeOf(&errorVar).Elem()

result := funcReturn[0].Interface()
possiblyError := funcReturn[len(funcReturn)-1].Interface()
if lastVal != nil && !reflect.TypeOf(possiblyError).AssignableTo(errorType) {
	// Last value is always the error if there is an error, if non-nil return it.
	writeHTTPErrorResponse(w, http.StatusInternalServerError, errorStatus, fmtFunctionError(lastVal))
} else if !reflect.TypeOf(result).AssignableTo(errorType) {
	// If the first value is not an error then it is the result, return it.
	returnVal, _ := json.Marshal(firstVal)
	fmt.Fprintf(w, string(returnVal))
}
```